gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/Dataset.java

```diff
@@ -0,0 +1,285 @@
+
+    private final BigQueryOptions options;
+    private Iterable<BaseTableInfo> infoIterable;
+    private transient BigQuery bigquery;
+
+    public LazyTableIterable(BigQueryOptions options, Iterable<BaseTableInfo> infoIterable) {
+      this.options = options;
+      this.infoIterable = infoIterable;
+      this.bigquery = options.service();
+    }
+
+    private void readObject(ObjectInputStream in) throws IOException, ClassNotFoundException {
+      in.defaultReadObject();
+      this.bigquery = options.service();
+    }
+
+    @Override
+    public Iterator<Table> iterator() {
+      return Iterators.transform(infoIterable.iterator(), new Function<BaseTableInfo, Table>() {
+        @Override
+        public Table apply(BaseTableInfo tableInfo) {
+          return new Table(bigquery, tableInfo);
+        }
+      });
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(options, infoIterable);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      if (!(obj instanceof LazyTableIterable)) {
+        return false;
+      }
+      LazyTableIterable other = (LazyTableIterable) obj;
+      return Objects.equals(options, other.options)
+          && Objects.equals(infoIterable, other.infoIterable);
+    }
+  }
+
+  /**
+   * Constructs a {@code Dataset} object for the provided {@code DatasetInfo}. The BigQuery service
+   * is used to issue requests.
+   *
+   * @param bigquery the BigQuery service used for issuing requests
+   * @param info dataset's info
+   */
+  public Dataset(BigQuery bigquery, DatasetInfo info) {
+    this.bigquery = checkNotNull(bigquery);
+    this.info = checkNotNull(info);
+  }
+
+  /**
+   * Creates a {@code Dataset} object for the provided dataset's user-defined id. Performs an RPC
+   * call to get the latest dataset information.
+   *
+   * @param bigquery the BigQuery service used for issuing requests
+   * @param dataset dataset's user-defined id
+   * @return the {@code Dataset} object or {@code null} if not found.
+   * @throws BigQueryException upon failure
+   */
+  public static Dataset load(BigQuery bigquery, String dataset) {
+    DatasetInfo info = bigquery.getDataset(dataset);
+    return info != null ? new Dataset(bigquery, info) : null;
+  }
+
+  /**
+   * Returns the dataset's information.
+   */
+  public DatasetInfo info() {
+    return info;
+  }
+
+  /**
+   * Checks if this dataset exists.
+   *
+   * @return {@code true} if this dataset exists, {@code false} otherwise.
+   * @throws BigQueryException upon failure
+   */
+  public boolean exists() {
+    return bigquery.getDataset(info.datasetId(), BigQuery.DatasetOption.fields()) != null;
+  }
+
+  /**
+   * Fetches current dataset's latest information.
+   *
+   * @param options dataset options
+   * @return a {@code Dataset} object with latest information.
+   * @throws BigQueryException upon failure
+   */
+  public Dataset reload(BigQuery.DatasetOption... options) {
```

Comment:
Just noticed that `Storage` has the same issue and that `ResourceManager` does not
check for `null` in `Project.load` (and therefore does not documented it - /cc @ajkannan).